### ISSUE TITLE
Query parsing compatibility issues epic

### ISSUE THREAD

**Issue by bbeaudreault** [youtube/vitess#2466]
I wrote a query analyzer and analyzed about 100TB of our mysql query logs in order to find all (or at least the vast majority) of the ways in which our existing queries aren't compatible with the vitess parser.  Here's what I came up with:

#### Waiting on PRs
None at the moment.
#### In Progress
None at the moment.
#### Still todo

- [ ] Multi-Table DELETE and UPDATE syntax `DELETE one, two FROM ... JOIN` https://github.com/youtube/vitess/issues/2439

- [ ] `JOIN ... USING(col)`, rather than `ON <expr>`

#### Deferred

~- [ ] `SAVEPOINT foo` and `ROLLBACK SAVEPOINT foo` - http://dev.mysql.com/doc/refman/5.7/en/savepoint.html. Put on the backburner for now; I realized that all of the usages were from xtrabackup~

- [ ] `LAST_INSERT_ID()` in `ON DUPLICATE KEY UPDATE`. For instance `INSERT .. ON DUPLICATE KEY UPDATE foo = "bar", id = LAST_INSERT_ID(id)`. We punted on this. It's only used by a small handful of apps, for us, and has some other consequences of existing. So we're going to try to kill those usages, and may revisit support in the future if that effort turns out to be unsuccessful.

#### Not a problem

For posterity, here I will include errors I originally reported but then on further investigation realized were not a problem

~- [x] `CASE WHEN ...` in right hand side of `INSERT ... ON DUPLICATE KEY UPDATE foo = CASE WHEN ...`. According to http://dev.mysql.com/doc/refman/5.7/en/insert.html, the right hand side is `col_name=expr`. Expr can be any of https://dev.mysql.com/doc/refman/5.7/en/expressions.html.~
~- [x] `UNHEX()` function~

Both of these were actually due to VALUES() func, which is handled above.

#### Done
- [x] `SELECT UNION ALL SELECT` - unsupported construct (so parsed, but not handled)

- [x] `INSERT <table>` - They support `INSERT INTO <table>`, but `INTO` is considered required.

- [x] `SELECT ... LIMIT <N> OFFSET <N>` - Vitess supports `LIMIT <offset>, <limit>`, but not the `OFFSET` keyword.

- [x] https://github.com/youtube/vitess/pull/2490 Double bar `||` instead of `OR` syntax. https://dev.mysql.com/doc/refman/5.7/en/expressions.html

- [x] https://github.com/youtube/vitess/pull/2497 `UNIX_TIMESTAMP()/UTC_TIMESTAMP/CURRENT_DATE/CURRENT_TIMESTAMP` functions, with and without parenthesis. Also `REPLACE('string', 'pattern', 'sub')` and `LEFT()` functions

- [x] https://github.com/youtube/vitess/pull/2502 Complex expressions in ORDER BY. I.E. `order by floor(floor(submission_timestamp / 1000 / 60) * 60 * 1000)`. This parses but is disallowed by Vitess: `Error: unsupported: complex expression in order by`.

- [x] https://github.com/youtube/vitess/pull/2479 `VALUES(foo)` function in `INSERT ... ON DUPLICATE KEY UPDATE` syntax http://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_values

- [x] https://github.com/youtube/vitess/pull/2505 `SELECT (foo OR bar) AS baz` -- fails on OR. There are also examples without the parenthesis. https://dev.mysql.com/doc/refman/5.7/en/expressions.html. Same as above, but probably more likely to be hard

- [x] https://github.com/youtube/vitess/pull/2505 Columns as boolean values, i.e. `SELECT foo, bar FROM x WHERE foo AND bar IS NULL`

- [x] https://github.com/youtube/vitess/pull/2505 Booleans as values, i.e. `INSERT INTO tbl (foo, bar) VALUES ("test", true)` and other variants, like `SET  foo = TRUE`, etc

- [x] https://github.com/youtube/vitess/pull/2505 boolean expression in `ORDER BY` clause. i.e. `ORDER BY foo IS NULL` https://dev.mysql.com/doc/refman/5.7/en/expressions.html. Might be easy if we can support boolean_expression just in ORDER BY as a stopgap until the full `column as boolean`/`boolean as value` support below.

- [x] https://github.com/youtube/vitess/pull/2505 String value in CASE statement, i.e. `IF(MAX(CASE status WHEN 'STARTED' THEN 1 ELSE 0 END) = 1, 'STARTED', 'ENDED') AS status. Vitess fails on 'STARTED', and succeeds if I replace that with a boolean value. But string values are supported. `https://dev.mysql.com/doc/refman/5.7/en/control-flow-functions.html#operator_case

- [x] https://github.com/youtube/vitess/pull/2483 `REPLACE INTO` - syntax is similar to `INSERT` but deletes old row if unique/primary key conflict http://dev.mysql.com/doc/refman/5.7/en/replace.html

- [x] https://github.com/youtube/vitess/pull/2500 Exclamation/bang `!` instead of `NOT` syntax. http://dev.mysql.com/doc/refman/5.7/en/logical-operators.html#operator_not

- [x] https://github.com/youtube/vitess/pull/2500 `foo DIV 30` arithmetic operator DIV (divide) http://dev.mysql.com/doc/refman/5.7/en/arithmetic-functions.html.

- [x] https://github.com/youtube/vitess/pull/2500 `CONVERT(foo, TYPE)` / `CAST(expr AS TYPE)`

- [x] https://github.com/youtube/vitess/pull/2500 table qualifier in UPDATE, i.e. `UPDATE foo SET foo.bar = "test"`.

- [x] https://github.com/youtube/vitess/pull/2500 SQL_CACHE/SQL_NO_CACHE

- [x] https://github.com/youtube/vitess/pull/2500 MATCH ... AGAINST

- [x] https://github.com/youtube/vitess/pull/2500 `GROUP_CONCAT()` function -- http://dev.mysql.com/doc/refman/5.7/en/group-by-functions.html#function_group-concat

- [x] https://github.com/youtube/vitess/pull/2500 BINARY operator https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#operator_binary

- [x] table alias in update e.g. `UPDATE foo bar SET bar.a = 5, bar.b = 6`
- [x] table qualifier in insert e.g. `insert into foo (foo.a, foo.b, foo.c) values (...)`

**Comment by bbeaudreault** [comment]
@sougou would love your feedback and/or guidance on this list.  I tried to group them by how hard I think they should be, but if you have any re-categorizations based on your experience that'd be helpful.  Also, looking for comments on the viability of some of the harder ones at the bottom of the list.

If you need more details on any of these let me know. This was the preliminary list that I wanted to get feedback on, and I will be trying to add more details and examples soon.

**Comment by bbeaudreault** [comment]
We also saw many instances of `symbol foo not found`. I think many of these are things we can fix in the queries themselves, because people did not explicitly specify the table for a column in situations where there are more than 1 table (joins). Not sure if there's anything we can do to make vitess a little more flexible on this front.

**Comment by bbeaudreault** [comment]
I submitted https://github.com/youtube/vitess/pull/2479 for supporting the VALUES() func

**Comment by bbeaudreault** [comment]
I submitted https://github.com/youtube/vitess/pull/2483 for supporting `REPLACE INTO`

**Comment by adkhare** [comment]
@bbeaudreault it would help if you can link the PR to the every line item in the epic. just a nice to have as it becomes easy to track

**Comment by bbeaudreault** [comment]
@adkhare I updated the list to make it easier to track. I'll keep it updates as things keep moving. Thanks!

**Comment by acharis** [comment]
DIV arithmetic operator in progress

**Comment by bbeaudreault** [comment]
Moved `JOIN ... USING(col)` to the bottom of the harder list. I hacked away at this for a while and couldn't make it work. The logic around pushing and merging routes is pretty complicated and it's unclear how to work Using into that. Will come back to it if no one else gets to it.

**Comment by acharis** [comment]
Exclamation/bang ! instead of NOT syntax. in progress

**Comment by bbeaudreault** [comment]
@adkhare are you still working on complex order by clauses? Should I move that to in-progress?

**Comment by bbeaudreault** [comment]
Added complex order by to the "Waiting on PRs" state.  Verified that @adkhare's PR solves our test case:

`select col from user where id = 1 order by floor(floor(id / 1000 / 60) * 60 * 1000)`

**Comment by acharis** [comment]
in progress:
table qualifier in update
group concat

**Comment by bbeaudreault** [comment]
Found another incompatibility, hopefully an easy one:

```
select * from t where match(a) against ('foo' in boolean mode)
```
https://dev.mysql.com/doc/refman/5.5/en/fulltext-search.html#function_match

**Comment by bbeaudreault** [comment]
Another easy one:

`SELECT SQL_NO_CACHE ....`

Parse test returns:

```
--- FAIL: TestValid (0.01s)
	parse_test.go:691: out: select SQL_NO_CACHE as a from t where b, want select SQL_NO_CACHE a from t where b
```

It seems to think it's a column alias

**Comment by acharis** [comment]
in progress:
SQL_CACHE and SQL_NO_CACHE select directives
match function

**Comment by bbeaudreault** [comment]
We're coming down home stretch. Did another run through with `vtparse` with all of our current fixes. Got a few more hits, which almost all seem easy:

- [ ] `&&` rather than `AND` in boolean expressions
- [ ] `DATE()` function
- [ ] table alias in update e.g. `UPDATE foo bar SET bar.a = 5, bar.b = 6`
- [ ] table qualifier in insert e.g. `insert into foo (foo.a, foo.b, foo.c) values (...)`
- [ ] Some people have used keywords as column names or aliases -- not sure if that's something we can handle, or need to fix their queries. I.E. `SELECT foo AS date WHERE language = 'en' or key = 'test` - date, key, and language are keywords, all throw a syntax exception 
- [ ] Seeing usages of charset introducers, i.e. `SELECT foo WHERE bar = _ascii 'test'` https://dev.mysql.com/doc/refman/5.7/en/charset-introducer.html
- [ ] `#` as a comment marker https://dev.mysql.com/doc/refman/5.7/en/comments.html
- [ ] `escape` keyword in LIKE expression: https://dev.mysql.com/doc/refman/5.7/en/string-comparison-functions.html#operator_like
- [ ] `insert into test values()` -- no values in the values list.  this table has only 1 auto_increment column in it

We're also seeing a couple order by issues:

```sql
-- count(*) in the order by is seen as a complex order by.
SELECT COUNT(*), foo FROM test GROUP BY foo HAVING COUNT(*) > 1 ORDER BY COUNT(*) DESC

-- column number 3 is out of bounds, for a table which has more than 3 columns
SELECT * FROM test ORDER BY 3 ASC
```

**Comment by bbeaudreault** [comment]
Submitted https://github.com/youtube/vitess/pull/2515 which handles the 6 checked off cases above.  @acharis I started looking at the table alias in update and table qualifier in insert, but decided it would conflict with your PR which is adding table qualifier in the RHS of update. So I'll let you handle those 2

**Comment by bbeaudreault** [comment]
We're going to punt on charset introducers for now. Removed the following task from the above list:

- [x] ~~Seeing usages of charset introducers, i.e. `SELECT foo WHERE bar = _ascii 'test'` https://dev.mysql.com/doc/refman/5.7/en/charset-introducer.html~~

It's only used by 1 database as far as I can tell, and was complicated to support in the parser due to shift/reduce conflicts. We will try to remove the 1 use-case, and revisit in the future if necessary.

**Comment by o1egl** [comment]
@bbeaudreault can you share your query analyzer? I also want to analyze queries we have.

**Comment by bbeaudreault** [comment]
@o1egl https://gist.github.com/bbeaudreault/729887a513bc0e54dcccb5cbdb93ed4b

I'll reiterate the disclaimer I included there: I've not run this in any form in a few months. It doesn't even compile right now, but I think due to import paths that have changed since I wrote it. I'm providing it mostly as a starting point rather than a working tool

**Comment by bbeaudreault** [comment]
There are still 2 issues left here, but closing this epic as we have stopped work on this for now. We will almost certainly come back to multi-update/delete and possibly `join...using`, and will create new issues for them when the time comes.

Thanks all for the work and help here; we made a huge list of query types compatible with Vitess as part of this.

**Comment by bbeaudreault** [comment]
With the recent work from sougou and I, the only thing left is `JOIN...USING`
